h5py/_hl/selections.py

```diff
@@ -175,8 +175,12 @@ def mshape(self):
         """ Shape of selection (always 1-D for this class) """
         return (self.nselect,)
 
-    def broadcast(self, target_shape):
-        """ Get an iterable for broadcasting """
+    def broadcast(self, target_shape, collective=False):
+        """ Get an iterable for broadcasting.
+
+        Setting `collective` includes zero length selections that are
+        required for collective operations.
```

Comment:
Actually, looking at it some more, does the selections code need to know about collective mode at all? Shouldn't it handle empty selections correctly regardless?

I appreciate the desire to not change anything for normal I/O. But if we can avoid special-casing it as far as possible, the code is cleaner and better tested.

Reply:
I agree that it would be best to avoid special cases. If you're amenable to that I'll try and go that route. Hopefully the tests will illuminate any unintended problems.

Reply:
Oh, I see. The obvious thing for a generator over an empty selection to do is to yield nothing. But that won't work in collective mode.

What happens if the different workers get different (non-zero) numbers of chunks from `.broadcast()`? Does that cause a hang because they're doing different numbers of HDF5 writes?

I feel like I'm missing something here. It shouldn't need this much code to be aware that it's doing something collectively. Maybe the answer is just to block broadcasting for collective I/O, so you have to set up the right shape of data yourself.

Reply:
> What happens if the different workers get different (non-zero) numbers of chunks from `.broadcast()`? Does that cause a hang because they're doing different numbers of HDF5 writes?

That is exactly what happens.

Empirically, just changing the clause from
```diff
-        if nchunks == 1 or (nchunks == 0 and collective):
+        if nchunks < 2:
```
does actually work fine. I also think it probably doesn't have any odd side effects because if `nchunks` was zero, then the `nselect == 0` check probably means that normal IO calls don't get there.

Reply:
@takluyver I'm starting to agree with your suggestion that we should forbid broadcasting in collective mode, or at least strongly restrict it. In theory you might want to do something like this:
```
arr = np.ones(10)
dset = fh.create_dataset('test', shape=(3, 10))
with dset.collective:
    if comm.rank == 0:
        dset[:2, :10] = arr   # 2 chunks, yields and writes twice on MPI rank=0
    else:
        dset[2:, :10] = arr  # 1 chunk, yields and writes once on MPI rank=1
```
which even with either of the changes I've tried would hang, because there's fundamentally a different number of chunks needed to be written by each. I think there's two options here:

1. Disallow broadcasting when doing MPI collective writes.
2. Before writing, communicate with all other ranks to determine how many chunks will be written, and then insert empty writes to keep the number the same across all ranks.

The latter seems like a bit of a hack.

Reply:
What happens if one MPI process throws an error? Because if we tried to disallow broadcasting, at least in the obvious way, I suspect rank 0 in your example would throw an error, while rank 1 would do its 1 chunk write and then wait for a matching write in rank 0 which isn't going to happen.

This is ugly. I'm trying to think about how best to deal with it, but I'm not sure I know enough about MPI.

I also notice that we should update the docs, because the section on collective I/O doesn't even mention with `with dset.collective` context manager:
http://docs.h5py.org/en/stable/mpi.html#collective-versus-independent-operations

Reply:
In the way I was thinking of implementing this, I don't think it would break. You probably want to insist that each ranks file and memory shapes match exactly. So rank=1 wouldn't work here because it tries to write a shape=(10,) array into a shape=(2,10) selection.

However, despite that I'd probably have each rank do it's preparation for writing, and then just before writing have all ranks communicate if they have any errors and have then *all* throw an exception about the need to broadcast. Something like:
```python

local_mpi_broadcast_error = (memshape != selshape)

# Reduce to see if *any* rank had an error
global_mpi_bcast_error = self.comm.allreduce(local_mpi_broadcast_error, op=MPI.MAX)

if local_mpi_broadcast_error:
    raise ValueError("Cannot broadcast in MPI collective mode (mem shape=%s, file selection=%s) % ....)

if global_broadcast_error:
     raise ValueError("Yikes. Another rank tried to broadcast during a write.")

# ... actually do the write now
```

A bit like that anyway. Let me change it again and send in a PR.

Reply:
Thinking about it more, I think we can do it well enough with a simple check, without exchanging messages, which means less to go wrong. It will always be possible to create a hang if you branch on `if comm.rank ...`, but we want to make it obvious that things like this are a problem:

```python
dset = fh.create_dataset('test', shape=(10,))
chunk_start = int(comm.rank / comm.size)
chunk_end = int((comm.rank + 1) / comm.size)
with dset.collective:
    dset[chunk_start:chunk_end] = comm.rank
```

At the moment, that will work (inefficiently: #1067) for 4 workers, but fail for 8, because some workers will write 12 values and some 13. But if we forced people to do the broadcasting with numpy beforehand, it would work:

```python
chunk = np.full(chunk_end - chunk_start, comm.rank)
with dset.collective:
    dset[chunk_start:chunk_end] = comm.rank
```

I think it needs to be a fairly strict check: not just that broadcast would only return one chunk, but that the dimensions of the data match those of the selection, even when those are length-1 dimensions.